redfish/provider/resource_redfish_storage_controller.go

```diff
@@ -0,0 +1,763 @@
+	var diags diag.Diagnostics
+
+	// get storage controller by using system id, storage id, controller id
+	system, storageController, err := getStorageControllerInstance(
+		service,
+		state.SystemID.ValueString(),
+		state.StorageID.ValueString(),
+		state.ControllerID.ValueString(),
+	)
+	if err != nil {
+		diags.AddError("Error when retrieving storage controller", err.Error())
+		return diags
+	}
+
+	storageControllerExtended, err := dell.StorageController(storageController)
+	if err != nil {
+		diags.AddError("Error when retrieving storage controller extended", err.Error())
+		return diags
+	}
+
+	diags = parseStorageControllerExtendedIntoState(ctx, storageControllerExtended, state)
+	if diags.HasError() {
+		return diags
+	}
+
+	diags = parseSecurityAttributesIntoState(ctx, storageControllerExtended, state)
+	if diags.HasError() {
+		return diags
+	}
+
+	state.ID = types.StringValue("redfish_storage_controller_resource")
+	state.SystemID = types.StringValue(system.ID)
+	return diags
+}
+
+// nolint: gocyclo, gocognit, revive
+func updateRedfishStorageController(ctx context.Context, service *gofish.Service, state, plan *models.StorageControllerResource) diag.Diagnostics {
+	var diags diag.Diagnostics
+
+	applyTime := plan.ApplyTime.ValueString()
+	resetType := plan.ResetType.ValueString()
+	resetTimeout := plan.ResetTimeout.ValueInt64()
+	jobTimeout := plan.JobTimeout.ValueInt64()
+
+	jobWait := true
+	if applyTime == string(redfishcommon.AtMaintenanceWindowStartApplyTime) ||
+		applyTime == string(redfishcommon.InMaintenanceWindowOnResetApplyTime) {
+		if plan.MaintenanceWindow == nil || plan.MaintenanceWindow.StartTime.IsUnknown() {
+			diags.AddError("Input param is not valid",
+				"Please set `maintenance_window` when `apply_time` is `AtMaintenanceWindowStart` or `InMaintenanceWindowOnReset`")
+			return diags
+		}
+		// when apply_time is AtMaintenanceWindowStart or InMaintenanceWindowOnReset, skip wait for job to finish
+		jobWait = false
+	}
+
+	// Lock the mutex to avoid race conditions with other resources
+	redfishMutexKV.Lock(plan.RedfishServer[0].Endpoint.ValueString())
+	defer redfishMutexKV.Unlock(plan.RedfishServer[0].Endpoint.ValueString())
+
+	// OnReset case
+	if applyTime == string(redfishcommon.OnResetApplyTime) {
+		// Reboot the server
+		pOp := powerOperator{ctx, service, plan.SystemID.ValueString()}
+		_, err := pOp.PowerOperation(resetType, resetTimeout, intervalStorageControllerJobCheckTime)
+		if err != nil {
+			diags.AddError(RedfishJobErrorMsg, err.Error())
+			return diags
+		}
+	}
+
+	isControllerModeAttributeChanged := storageControllerAttributesChanged(ctx, plan, state, true)
+	isAnyOtherStorageControllerAttributeChanged := storageControllerAttributesChanged(ctx, plan, state, false)
+	isAnySecurityAttributeChanged := securityAttributesChanged(ctx, plan, state)
+
+	var jobURL string
+	if isControllerModeAttributeChanged {
+		if isAnyOtherStorageControllerAttributeChanged || isAnySecurityAttributeChanged {
```

Comment:
hi @MangirishK , checked at cases, not find the mode change case. 
We have discussed one point when the mode is changed, we need to make sure the flag should be set to disable.(if i misunderstand, please correct me)